### DISCUSSION TITLE
Breaking change with update-snapshot property introduction

### DISCUSSION THREAD

**Discussion by lanwen** [origin-energy/java-snapshot-testing#151] [General]
Hello,

I updated recently from 4.0.2 to 4.0.3 and immediately got a failure on CI due to snapshot tests, which I suppose shouldn't be the case for the patch release when nothing changed except the dependency.

Tests started to complain on 

```
au.com.origin.snapshots.exceptions.SnapshotExtensionException: isCI=true & update-snapshot=Optional[false]. Updating snapshots on CI is not allowed
```

as we specify `updateSnapshot=false` in our gradle config by default

I think the reason for that is the logic introduced with this change here: https://github.com/origin-energy/java-snapshot-testing/commit/194f256d8be66fa233b9cd9fb510f7ca03277922#diff-6c0dc58e8f7f6174ee8150ac6b242ba7bbebecf5ed998c9f786cf056163aa189R122
(#145)

It checks only for the presence of the property, but not the value, whereas before absence and the `false` value were equal.

Perhaps, you could bring the `false` value for the old property to be considered as `none`? That would be much appreciated to allow the smooth migration process.

p.s. the new flow with the property value change seems not obvious to me - would you mind describing the way you see it? As I'm always afraid to accidentally commit the property file after changing it to `all`. Thank you in advance.

**Comment by jackmatt2** [comment]
Yes - you're correct in that this change should not have broken anything in 4.X release.  I'll need to look into it.

Regarding the flow - I see it like this
- By default, the CI env var will not allow updating snapshots.  It ignores what is defined in `snapshot.properties`
- It should not be necessary to explicitly define it as false for CI environments

If you accidentally checkin the change to`snapshot.properties` you are protected in two ways
- CI will ignore what you specified in `snapshot.properties` and still fail the tests
- PR reviews and identify the mistaken change to the file

**Comment by uk-taniyama** [comment]
It would be better to have external control over the updateSnapshot.
It is easier to control if the environment variable -> snapshot.properties.

**Comment by jackmatt2** [comment]
@uk-taniyama it's currently not an environment variable - it's a system property.  An environment variable would make it easier for people to control.  But I'm not too enthusiastic to now support 3 different ways of accomplishing this.

**Reply by uk-taniyama** [reply]
Thanks for your reply.
I will use MySnapshotConfig.class like this.

```
import java.util.Optional;

import au.com.origin.snapshots.SnapshotProperties;
import au.com.origin.snapshots.config.PropertyResolvingSnapshotConfig;
import au.com.origin.snapshots.exceptions.MissingSnapshotPropertiesKeyException;

public class MySnapshotConfig extends PropertyResolvingSnapshotConfig {
    private static final Optional<String> all = Optional.of("");
    private static final Optional<String> none = Optional.empty();

    @Override
    public Optional<String> updateSnapshot() {
        if (isCI()) {
            return none;
        }

        try {
            String updateSnapshotEnvVar = SnapshotProperties.getOrThrow("update-snapshot-env-var");
            if (updateSnapshotEnvVar != null && !updateSnapshotEnvVar.isEmpty()) {
                String updateString = System.getenv(updateSnapshotEnvVar);
                if (updateString == null || updateString.equals("none")) {
                    return none;
                }
                if (updateString.equals("all")) {
                    return all;
                }
                return Optional.of(updateString);
            }
        } catch (MissingSnapshotPropertiesKeyException ignore) {
        }
        return super.updateSnapshot();
    }
}

```

**Comment by jackmatt2** [comment]
I have put in a fix for this in 4.0.5

I'll now move this to a discussion

**Comment by muehmar** [comment]
Actually, I was pretty happy with the old way to update snapshots by passing -PupdateSnapshot to the test task. Sure, it was necessary to use some gradle tricks to get it working, but once it worked, it was pretty convenient. Personally, I dont really like the way it is implemented now with setting the property in the file to `all` and setting it back again to `none`. Is there a way you could think of to support as alternative to modify files to update snapshots? Or perhaps still support the old way (I know its still supported but I suspect you will remove it with the next major version)?

**Reply by grimsa** [reply]
I also think that being able to control snapshot updating without requiring file `snapshot.properties` file modification is important.

For example, during development it's convenient to sometimes force update of all snapshots (e.g. after making some change that affects many snapshot files). In that case I can think of two methods:
1. Open `snapshot.properties`, change `update-snapshot` to `all`, run tests, revert the change in `snapshot.properties`
2. Add a custom run configuration in your IDE so you can run it without having to make any changes to any files or remember property names.

IMO method 2 is faster.

The current implementation of method 2 looks like this:

In `build.gradle`:
```
test {
    useJUnitPlatform()
    if (project.hasProperty('forceUpdateTestSnapshots')) {
        // Note: this JVM property configures java-snapshot-testing library to overwrite snapshots
        systemProperty 'updateSnapshot', 'true'
    }
}
```
and then it can be run using `gradle test -PforceUpdateTestSnapshots`, which can be saved as a persistent IDE run configuration.

It does result in a lot of deprecation warnings though:
```
[WARN ] 2025-08-28 15:03:20,900 [Test worker] a.c.o.s.config.PropertyResolvingSnapshotConfig - 

** Deprecation Warning **
This feature will be removed in version 5.X
Passing -PupdateSnapshot will be removed in a future release. Consider using snapshot.properties 'update-snapshot' toggle instead
```

It should likely be possible to configure Gradle to modify the `snapshot.properties` file before test run, and then revert it after test run completes, but it would be more complex + more fragile (if you stop Gradle mid-test-run, the file would remain modified).

**Reply by jackmatt2** [reply]
I can probably maintain both approaches I guess.  Env var will take precedence over snapshot.properties.
